### PR TITLE
feat: Tap 4 Implementation (TUFClient interface)

### LINKED ISSUE #348: Implement TAP 4: Multi-repository mappings
This tracking issue is to implement [TAP 4](https://github.com/theupdateframework/taps/blob/master/tap4.md): Multiple repository consensus on entrusted targets.

To implement this, I propose creating a NEW type `MultiRepositoryClient` which takes a `map.json` file. It would expose `NewClient`, `Init`, `Update`, `Targets`, and  `Download` similar to the existing client.

Internally, it would contain a map from repository name to existing TUF client updaters. The logic for priority and consensus would exist in the `MultiRepositoryClient`.

The purpose of this is to use in Sigstore's TUF client to support multiple repositories, augmenting the public root with custom, hosted, or test repositories with an AND relationship.

cc @kommendorkapten @haydentherapper

### PR DESCRIPTION
**Pull request by dibrinsofor**
See #348 

Release Notes: 

**Types of changes**:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). **Please ensure that your PR title** is a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) breaking change (with a `!`, as in `feat!: change foo`). 

**Description of the changes being introduced by the pull request**:  Began abstracting Client into interface in interface.go for future use towards multi-client mapping. currently setup with Update, Download, VerifyDigest, Target and Targets functionality. with @znewman01 and @vaikas

**Please verify and check that the pull request fulfills the following requirements**:

- [X] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by dibrinsofor** [comment]
I had some issues rebasing with just these commits as I had already started to implement some of your suggestions locally, hence the new PR.

File: client/interface.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	// dest will be deleted and an error returned in the following situations:
+	//
+	//   - The target does not exist in the local targets.json
+	//   - Failed to fetch the chain of delegations accessible from local snapshot.json
+	//   - The target does not exist in any targets
+	//   - Metadata cannot be generated for the downloaded data
+	//   - Generated metadata does not match local metadata for the given file
+	Download(name string, dest Destination) (err error)
+
+	VerifyDigest(digest string, digestAlg string, length int64, path string) error
```
Comment (asraa) [inline client/interface.go:21]: why does this need to be exposed? Is this for verifying a target against some local metadata?

**Review by asraa — APPROVED** [review]
other than the usage of verifydigest, lgtm

File: client/interface.go (hunk 1 above)
Comment (znewman01) [reply client/interface.go:21]: > why does this need to be exposed? 

The unhelpful answer is that it's part of the effective public interface that a Client exposes and it's covered in `client_test.go` so if we want the whole test suite to pass we need to expose it 😛 

I think specifically it's for OCI and other, similar contexts: https://github.com/theupdateframework/go-tuf/pull/170

It feels like a useful part of the public interface generally IMO. It's already part of the `Client` struct so we're somewhat committed to it. I think if we have this method in `Client`, it should be part of the `TUFClient` interface. I

You can implement this by calling `Target(path)` and then checking the hash etc. manually; that seems slightly error-prone 🤷 I'd consider removing it from the struct and the interface, can't find anybody that uses it. But I think let's make that a separate discussion.

>  Is this for verifying a target against some local metadata?

Roughly, yes. @dibrinsofor could we add a comment like the following?

> VerifyDigest verifies a target at the given path by its digest.
>
> This is useful in contexts where the targets are distributed via some channel besides TUF (e.g., OCI registries).

Comment (asraa) [reply client/interface.go:21]: Got it, the linked PR was useful for context!

just for future reference testing in the same package doesn't require only using exported methods, you can test internal methods in the same package.

I do think it's useful, but I almost think my gripe was that the interface was kind of confusing. I'm supplying a hash, then alg, then size, and then path. Presumably of my target, and expect that this method will get the target the path and compare the metadata. 

I think VerifyTargetWithDigest(targetPath, digestAlg, digest, length) woulda made more sense, but we can work on it separately from this interface discusison

**Review by asraa — APPROVED** [review]

**Review by znewman01 — APPROVED** [review]

**Comment by znewman01** [comment]
This looks good expect for two things:

1. You need to add the [DCO](https://developercertificate.org/) to your commits: https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md
2. Can you modify the PR to remove "Fixes #348"? It should just say "See #348" or something

**Review by znewman01 — DISMISSED** [review]

**Comment by znewman01** [comment]
Sorry about this @dibrinsofor but can you rebase? We can do a quick re-review, should be an easy change then.

**Comment by dibrinsofor** [comment]
@znewman01 all done!

**Review by znewman01 — APPROVED** [review]
CC @asraa for an extra maintainer approval

**Pull request closed without merging**